src/Features/CSharp/Portable/Snippets/ConsoleSnippetProvider.cs

```diff
@@ -0,0 +1,179 @@
+            if (syntaxFacts.IsInNonUserCode(syntaxTree, position, cancellationToken))
+            {
+                return false;
+            }
+
+            var semanticModel = await document.ReuseExistingSpeculativeModelAsync(position, cancellationToken).ConfigureAwait(false);
+            var syntaxContext = (CSharpSyntaxContext)document.GetRequiredLanguageService<ISyntaxContextService>().CreateContext(document, semanticModel, position, cancellationToken);
+
+            if (!ShouldDisplaySnippet(syntaxContext))
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        private static bool ShouldDisplaySnippet(CSharpSyntaxContext context)
+        {
+            var token = context.LeftToken;
+            var isDirectlyInUndesirableLocation = token.GetAncestors<SyntaxNode>()
+                .Any(node => node.IsKind(SyntaxKind.ParameterList) ||
+                             node.IsKind(SyntaxKind.SimpleLambdaExpression) ||
+                             node.IsKind(SyntaxKind.ArgumentList) ||
+                             node.IsKind(SyntaxKind.RecordDeclaration) ||
+                             node.IsKind(SyntaxKind.ObjectCreationExpression) ||
+                             node.IsKind(SyntaxKind.SwitchExpression));
+
+            if (isDirectlyInUndesirableLocation)
+            {
+                return false;
+            }
+
+            var isExpressionInVariable = token.GetAncestors<SyntaxNode>()
+                .Any(node => node.IsKind(SyntaxKind.ParenthesizedLambdaExpression) || node.IsKind(SyntaxKind.AnonymousMethodExpression));
+            var isInVariableDeclaration = token.GetAncestors<SyntaxNode>().Any(node => node.IsKind(SyntaxKind.VariableDeclaration));
+
+            if (isInVariableDeclaration && !isExpressionInVariable)
+            {
+                return false;
+            }
+
+            var isInsideMethod = token.GetAncestors<SyntaxNode>()
+               .Any(node => node.IsKind(SyntaxKind.MethodDeclaration) ||
+                            node.IsKind(SyntaxKind.ConstructorDeclaration) ||
+                            node.IsKind(SyntaxKind.LocalFunctionStatement) ||
+                            node.IsKind(SyntaxKind.AnonymousMethodExpression) ||
+                            node.IsKind(SyntaxKind.ParenthesizedLambdaExpression));
+
+            var isInNamespace = token.GetAncestors<SyntaxNode>()
+                .Any(node => node.IsKind(SyntaxKind.NamespaceDeclaration) ||
+                             node.IsKind(SyntaxKind.FileScopedNamespaceDeclaration));
+
+            if (isInNamespace && !isInsideMethod)
+            {
+                return false;
+            }
+
+            if (!isInsideMethod && !context.IsGlobalStatementContext)
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        protected override string GetSnippetDisplayName()
```

Comment:
Nit: Feels to me like a `DisplayName` property would be better. That would move it up to the top of a file, so when browsing through various snippets it would be more apparent